Handlers - Inventory list

rwx-yxu opened the pull request:
Contains the initial handler functions for an inventory list object. 
Supported methods added:

- GET /inventorylist/{id} - Returns a inventory list object
- POST /inventorylist/ - Creates a inventory list object. Returns the created at value in json
- DELETE /inventorylist/{id} - DELETES a inventory list object
- PUT /inventorylist/{id} - Updates a inventory list. Returns the updated at value in json

rwx-yxu committed:
Converting template to dockerfile. Handlers are now handled separately.

- I had to change the template to dockerfile because the go-middleware template was causing issues when I tried to implement chi router since it has the in built handler func. I want full control of the server to register the routes and connect to the database.
- I have added a main.go which was not possible before.
- Secret has been added for the database string connection for added security.

rwx-yxu committed:
Remove uneccessary comment during debug

rwx-yxu committed:
Update the postgres connection string name

Stack.yml also now includes the secret definition

rwx-yxu committed:
Use github registry for build workflow

rwx-yxu committed:
Update first migration to use unix timestamps rather than date time.

Converting the datatype for the datetimes to big int to store unix timestamp. This is because a timestamp would be more flexible than a date time string through out the codebase due to parsing the values to specific formats from go and postgres and then in vuejs when I eventually get to it.

The api should just hand over the time stamp and the eventual front end should handle the view for that to the user.

rwx-yxu committed:
JSON parse and validation added.

Adding JSON struct tags to InventoryList struct. Custom JSON validation package to handle JSON to struct validation. I added an additional validation check to throw an error if the id field is included as part of the JSON post.

rwx-yxu committed:
Post handler implementation to create inventory list

New post route added to create inventory list. Function should decode JSON, validate the struct, create list and then encode the struct to json as the response.

rwx-yxu committed:
Update inventory list added

For this update. This changes the updated at type to not null. The struct values for updated and created at also have json tags so they can be encoded.

Migrations have been updated to have the updated at field to be not null. Will default to 0.

WIP:
I decided to add a response struct in the handler to only allow for certain values to be returned in json. I figure that I dont need to return the entire json model back. The new piece of data will be the updated at value. The front end will need to check for the http response and get this new value.

rwx-yxu committed:
Implement inventory list created response struct

Response struct added to only expose the created at and id fields

rwx-yxu committed:
Inventory list delete handler added.

rwx-yxu committed:
Update to readme.

Adding preliminary information to readme.

Quasaer commented:
I just noticed that I'll need to add a new migration file to change the column types for the created at and updated dates to set them to big int and not null.

rwx-yxu committed:
Revert first migration up file.

Revert migration file to main because the migrations will not be applied. Should be part of a separate migration file.

rwx-yxu committed:
New migration step to alter the column types for date

This migration step will alter the column types for dates from timestamp to BIGINT to store the unix time.

rwx-yxu committed:
Correct the postgres syntax for alter column

Build failed due to syntax errors for migration. I needed to add type key word for the alter column and the set key word to add the not null contraint.

rwx-yxu committed:
Separate out the column alter and the not null constraint check

I think that you cannot make a data type change and add a not null constraint in the same alter column statement.

rwx-yxu committed:
Created at already has not null so is failing

Removed the not null constraint set for created at for both tables.

Quasaer commented:
Failing again. I noticed a error in the initial migration... The created_at column for inventory list is mapped to the 'NULL' constraint which is most likely why it is not able to change. I can either drop the column and remake or remove that constraint some how.

rwx-yxu committed:
FIX: Migration file updated to drop appropriate columns

The time stamp columns for inventory_list and inventory_items are now dropped and recreated. I had issues with trying to directly change the data type.

Quasaer commented:
I made the migrate down file the same because the reverse step will fail if it tried to set the big int type to timestamp. Not too big of a deal since the first downstep is to drop the tables anyways.
